File: go1_commerce/go1_commerce/page/commerce_dashboard/commerce_dashboard.js
frappe.pages['commerce-dashboard'].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'Commerce Dashboard',
		single_column: true
	});

	page.main.addClass('frappe-card');
	$(page.body).html('<div class="commerce-dashboard-container" style="padding: 15px;"></div>');
	new CommerceDashboard(page);
};

class CommerceDashboard {
	constructor(page) {
		this.page = page;
		this.container = $(page.body).find('.commerce-dashboard-container');
		this.render_loading();
		this.fetch_data();
	}

	render_loading() {
		this.container.html(`
			<div style="text-align: center; padding: 60px 0;">
				<div class="text-muted">${__("Loading dashboard...")}</div>
			</div>
		`);
	}

	fetch_data() {
		frappe.call({
			method: 'go1_commerce.go1_commerce.page.commerce_dashboard.commerce_dashboard.get_dashboard_data',
			callback: (r) => {
				if (r.message) {
					this.data = r.message;
					this.render();
				}
			}
		});
	}

	render() {
		let d = this.data;
		let nc = d.number_cards;
		let mc = d.monthly_comparison;

		let order_change = mc.previous_orders > 0
			? (((mc.current_orders - mc.previous_orders) / mc.previous_orders) * 100).toFixed(1)
			: 0;
		let revenue_change = mc.previous_revenue > 0
			? (((mc.current_revenue - mc.previous_revenue) / mc.previous_revenue) * 100).toFixed(1)
			: 0;

		this.container.html(`
			<!-- Number Cards Row -->
			<div class="row" style="margin-bottom: 20px;">
				${this.number_card("Total Orders", format_number(nc.total_orders), "blue", "shopping-cart")}
				${this.number_card("Total Revenue", format_currency(nc.total_revenue), "green", "dollar-sign")}
				${this.number_card("This Month Orders", format_number(nc.monthly_orders),
					parseFloat(order_change) >= 0 ? "green" : "red",
					"trending-up",
					order_change + "% vs last month"
				)}
				${this.number_card("This Month Revenue", format_currency(nc.monthly_revenue),
					parseFloat(revenue_change) >= 0 ? "green" : "red",
					"bar-chart-2",
					revenue_change + "% vs last month"
				)}
			</div>
			<div class="row" style="margin-bottom: 20px;">
				${this.number_card("Pending Orders", format_number(nc.pending_orders), "orange", "clock")}
				${this.number_card("Total Customers", format_number(nc.total_customers), "purple", "users")}
				${this.number_card("Avg Order Value", format_currency(nc.avg_order_value), "cyan", "activity")}
				<div class="col-sm-3">
					<div style="background: var(--card-bg); border-radius: 8px; padding: 20px; border: 1px solid var(--border-color); height: 100%;">
						<div style="font-size: 12px; color: var(--text-muted); text-transform: uppercase; letter-spacing: 0.5px;">Quick Links</div>
						<div style="margin-top: 10px;">
							<a href="/app/order" class="btn btn-xs btn-default" style="margin: 2px;">Orders</a>
							<a href="/app/product" class="btn btn-xs btn-default" style="margin: 2px;">Products</a>
							<a href="/app/customers" class="btn btn-xs btn-default" style="margin: 2px;">Customers</a>
						</div>
					</div>
				</div>
			</div>

			<!-- Charts Row 1: Sales Trend + Order Status -->
			<div class="row" style="margin-bottom: 20px;">
				<div class="col-sm-8">
					<div class="dashboard-chart-container" style="background: var(--card-bg); border-radius: 8px; padding: 20px; border: 1px solid var(--border-color);">
						<h6 style="margin-bottom: 15px; font-weight: 600;">Sales Trend (Last 12 Months)</h6>
						<div id="sales-trend-chart"></div>
					</div>
				</div>
				<div class="col-sm-4">
					<div class="dashboard-chart-container" style="background: var(--card-bg); border-radius: 8px; padding: 20px; border: 1px solid var(--border-color);">
						<h6 style="margin-bottom: 15px; font-weight: 600;">Order Status</h6>
						<div id="order-status-chart"></div>
					</div>
				</div>
			</div>

			<!-- Charts Row 2: Top Products + Payment Status -->
			<div class="row" style="margin-bottom: 20px;">
				<div class="col-sm-8">
					<div class="dashboard-chart-container" style="background: var(--card-bg); border-radius: 8px; padding: 20px; border: 1px solid var(--border-color);">
						<h6 style="margin-bottom: 15px; font-weight: 600;">Top Selling Products</h6>
						<div id="top-products-chart"></div>
					</div>
				</div>
				<div class="col-sm-4">
					<div class="dashboard-chart-container" style="background: var(--card-bg); border-radius: 8px; padding: 20px; border: 1px solid var(--border-color);">
						<h6 style="margin-bottom: 15px; font-weight: 600;">Payment Status</h6>
						<div id="payment-status-chart"></div>
					</div>
				</div>
			</div>

			<!-- Charts Row 3: Top Customers -->
			<div class="row" style="margin-bottom: 20px;">
				<div class="col-sm-6">
					<div class="dashboard-chart-container" style="background: var(--card-bg); border-radius: 8px; padding: 20px; border: 1px solid var(--border-color);">
						<h6 style="margin-bottom: 15px; font-weight: 600;">Top Customers by Revenue</h6>
						<div id="top-customers-chart"></div>
					</div>
				</div>
				<div class="col-sm-6">
					<div class="dashboard-chart-container" style="background: var(--card-bg); border-radius: 8px; padding: 20px; border: 1px solid var(--border-color);">
						<h6 style="margin-bottom: 15px; font-weight: 600;">Recent Orders</h6>
						<div id="recent-orders-table">${this.render_recent_orders(d.recent_orders)}</div>
					</div>
				</div>
			</div>
		`);

		this.render_charts();
	}

	number_card(title, value, color, icon, subtitle) {
		let color_map = {
			blue: "#4E79A7", green: "#59A14F", orange: "#F28E2B",
			red: "#E15759", purple: "#B07AA1", cyan: "#76B7B2"
		};
		let bg = color_map[color] || "#4E79A7";
		return `
			<div class="col-sm-3" style="margin-bottom: 10px;">
				<div style="background: ${bg}; border-radius: 8px; padding: 20px; color: #fff; min-height: 100px;">
					<div style="font-size: 12px; opacity: 0.85; text-transform: uppercase; letter-spacing: 0.5px;">${title}</div>
					<div style="font-size: 24px; font-weight: 700; margin-top: 8px;">${value}</div>
					${subtitle ? `<div style="font-size: 11px; opacity: 0.75; margin-top: 4px;">${subtitle}</div>` : ''}
				</div>
			</div>
		`;
	}

	render_recent_orders(orders) {
		if (!orders || !orders.length) return '<div class="text-muted text-center" style="padding: 30px;">No recent orders</div>';
		let rows = orders.map(o => {
			let status_color = {Placed: "blue", Processing: "orange", Completed: "green", Cancelled: "red", Delivered: "green"}[o.status] || "grey";
			return `<tr>
				<td><a href="/app/order/${o.name}">${o.name}</a></td>
				<td>${o.customer_name || ''}</td>
				<td><span class="indicator-pill ${status_color}">${o.status}</span></td>
				<td style="text-align:right;">${format_currency(o.total_amount)}</td>
			</tr>`;
		}).join('');
		return `<table class="table table-sm" style="font-size: 12px; margin: 0;">
			<thead><tr><th>Order</th><th>Customer</th><th>Status</th><th style="text-align:right;">Amount</th></tr></thead>
			<tbody>${rows}</tbody>
		</table>`;
	}

	render_charts() {
		let d = this.data;

		// Sales Trend - Line Chart
		if (d.sales_trend.labels.length) {
			new frappe.Chart("#sales-trend-chart", {
				type: "axis-mixed",
				height: 280,
				colors: ["#4E79A7", "#59A14F"],
				data: {
					labels: d.sales_trend.labels,
					datasets: [
						{name: "Orders", type: "bar", values: d.sales_trend.orders},
						{name: "Revenue", type: "line", values: d.sales_trend.revenue}
					]
				},
				tooltipOptions: {
					formatTooltipY: (v) => format_currency(v)
				}
			});
		}

		// Order Status - Pie Chart
		if (d.order_status.labels.length) {
			new frappe.Chart("#order-status-chart", {
				type: "pie",
				height: 280,
				colors: ["#4E79A7", "#F28E2B", "#59A14F", "#E15759", "#76B7B2", "#EDC948", "#B07AA1"],
				data: {
					labels: d.order_status.labels,
					datasets: [{values: d.order_status.values}]
				}
			});
		}

		// Top Products - Bar Chart
		if (d.top_products.labels.length) {
			new frappe.Chart("#top-products-chart", {
				type: "bar",
				height: 280,
				colors: ["#FF6F61"],
				data: {
					labels: d.top_products.labels,
					datasets: [{name: "Qty Sold", values: d.top_products.qty}]
				},
				barOptions: {spaceRatio: 0.4}
			});
		}

		// Payment Status - Pie Chart
		if (d.payment_status.labels.length) {
			new frappe.Chart("#payment-status-chart", {
				type: "percentage",
				height: 280,
				colors: ["#59A14F", "#F28E2B", "#E15759", "#76B7B2"],
				data: {
					labels: d.payment_status.labels,
					datasets: [{values: d.payment_status.values}]
				}
			});
		}

		// Top Customers - Bar Chart
		if (d.top_customers.labels.length) {
			new frappe.Chart("#top-customers-chart", {
				type: "bar",
				height: 280,
				colors: ["#36A2EB"],
				data: {
					labels: d.top_customers.labels,
					datasets: [{name: "Revenue", values: d.top_customers.spent}]
				},
				tooltipOptions: {
					formatTooltipY: (v) => format_currency(v)
				},
				barOptions: {spaceRatio: 0.4}
			});
		}
	}
}

File: go1_commerce/go1_commerce/page/commerce_dashboard/commerce_dashboard.py
import frappe
from frappe import _
from frappe.query_builder import DocType
from frappe.query_builder.functions import Count, Sum, Function
from frappe.utils import getdate, add_months, nowdate


@frappe.whitelist()
def get_dashboard_data():
	"""Get all dashboard data in a single call."""
	return {
		"number_cards": get_number_cards(),
		"sales_trend": get_sales_trend(),
		"order_status": get_order_status_distribution(),
		"payment_status": get_payment_status_distribution(),
		"top_products": get_top_products(),
		"top_customers": get_top_customers(),
		"recent_orders": get_recent_orders(),
		"monthly_comparison": get_monthly_comparison(),
	}


def get_number_cards():
	"""Get key metrics for number cards."""
	OrderDT = DocType("Order")
	today = getdate(nowdate())
	month_start = today.replace(day=1)

	total_orders = (
		frappe.qb.from_(OrderDT)
		.select(Count("*").as_("count"))
		.where(OrderDT.docstatus == 1)
		.run(as_list=True)
	)[0][0] or 0

	total_revenue = (
		frappe.qb.from_(OrderDT)
		.select(Sum(OrderDT.total_amount).as_("total"))
		.where(OrderDT.docstatus == 1)
		.where(OrderDT.payment_status == "Paid")
		.run(as_list=True)
	)[0][0] or 0

	monthly_orders = (
		frappe.qb.from_(OrderDT)
		.select(Count("*").as_("count"))
		.where(OrderDT.docstatus == 1)
		.where(OrderDT.order_date >= month_start)
		.run(as_list=True)
	)[0][0] or 0

	monthly_revenue = (
		frappe.qb.from_(OrderDT)
		.select(Sum(OrderDT.total_amount).as_("total"))
		.where(OrderDT.docstatus == 1)
		.where(OrderDT.payment_status == "Paid")
		.where(OrderDT.order_date >= month_start)
		.run(as_list=True)
	)[0][0] or 0

	pending_orders = (
		frappe.qb.from_(OrderDT)
		.select(Count("*").as_("count"))
		.where(OrderDT.docstatus == 1)
		.where(OrderDT.status.isin(["Placed", "Processing"]))
		.run(as_list=True)
	)[0][0] or 0

	CustomerDT = DocType("Customers")
	total_customers = (
		frappe.qb.from_(CustomerDT)
		.select(Count("*").as_("count"))
		.where(CustomerDT.customer_status == "Approved")
		.run(as_list=True)
	)[0][0] or 0

	avg_order_value = round(float(total_revenue) / max(total_orders, 1), 2)

	return {
		"total_orders": int(total_orders),
		"total_revenue": float(total_revenue),
		"monthly_orders": int(monthly_orders),
		"monthly_revenue": float(monthly_revenue),
		"pending_orders": int(pending_orders),
		"total_customers": int(total_customers),
		"avg_order_value": avg_order_value,
	}


def get_sales_trend():
	"""Get last 12 months sales trend."""
	OrderDT = DocType("Order")
	query = (
		frappe.qb.from_(OrderDT)
		.select(
			Function("DATE_FORMAT", OrderDT.order_date, "%Y-%m").as_("month"),
			Count("*").as_("orders"),
			Sum(OrderDT.total_amount).as_("revenue")
		)
		.where(OrderDT.docstatus == 1)
		.where(OrderDT.order_date >= add_months(nowdate(), -12))
		.groupby(Function("DATE_FORMAT", OrderDT.order_date, "%Y-%m"))
		.orderby(Function("DATE_FORMAT", OrderDT.order_date, "%Y-%m"))
	)
	data = query.run(as_dict=True)
	return {
		"labels": [d.month for d in data],
		"orders": [int(d.orders or 0) for d in data],
		"revenue": [float(d.revenue or 0) for d in data],
	}


def get_order_status_distribution():
	"""Get order count by status."""
	OrderDT = DocType("Order")
	query = (
		frappe.qb.from_(OrderDT)
		.select(OrderDT.status, Count("*").as_("count"))
		.where(OrderDT.docstatus == 1)
		.groupby(OrderDT.status)
		.orderby(Count("*"), order=frappe.qb.desc)
	)
	data = query.run(as_dict=True)
	return {
		"labels": [d.status for d in data],
		"values": [int(d.count) for d in data],
	}


def get_payment_status_distribution():
	"""Get order count by payment status."""
	OrderDT = DocType("Order")
	query = (
		frappe.qb.from_(OrderDT)
		.select(OrderDT.payment_status, Count("*").as_("count"))
		.where(OrderDT.docstatus == 1)
		.groupby(OrderDT.payment_status)
		.orderby(Count("*"), order=frappe.qb.desc)
	)
	data = query.run(as_dict=True)
	return {
		"labels": [d.payment_status for d in data],
		"values": [int(d.count) for d in data],
	}


def get_top_products():
	"""Get top 10 selling products."""
	Product = DocType("Product")
	OrderItem = DocType("Order Item")
	OrderDT = DocType("Order")
	query = (
		frappe.qb.from_(OrderItem)
		.inner_join(OrderDT).on(OrderDT.name == OrderItem.parent)
		.inner_join(Product).on(Product.name == OrderItem.item)
		.select(
			Product.item.as_("product_name"),
			Sum(OrderItem.quantity).as_("qty"),
			Sum(OrderItem.amount).as_("revenue")
		)
		.where(OrderDT.docstatus == 1)
		.where(OrderDT.payment_status == "Paid")
		.groupby(Product.name)
		.orderby(Sum(OrderItem.quantity), order=frappe.qb.desc)
		.limit(10)
	)
	data = query.run(as_dict=True)
	return {
		"labels": [d.product_name for d in data],
		"qty": [int(d.qty or 0) for d in data],
		"revenue": [float(d.revenue or 0) for d in data],
	}


def get_top_customers():
	"""Get top 10 customers by spend."""
	OrderDT = DocType("Order")
	query = (
		frappe.qb.from_(OrderDT)
		.select(
			OrderDT.customer_name,
			Count("*").as_("orders"),
			Sum(OrderDT.total_amount).as_("total_spent")
		)
		.where(OrderDT.docstatus == 1)
		.where(OrderDT.payment_status == "Paid")
		.groupby(OrderDT.customer_name)
		.orderby(Sum(OrderDT.total_amount), order=frappe.qb.desc)
		.limit(10)
	)
	data = query.run(as_dict=True)
	return {
		"labels": [d.customer_name for d in data],
		"orders": [int(d.orders or 0) for d in data],
		"spent": [float(d.total_spent or 0) for d in data],
	}


def get_recent_orders():
	"""Get 10 most recent orders."""
	OrderDT = DocType("Order")
	query = (
		frappe.qb.from_(OrderDT)
		.select(
			OrderDT.name,
			OrderDT.order_date,
			OrderDT.customer_name,
			OrderDT.status,
			OrderDT.payment_status,
			OrderDT.total_amount
		)
		.where(OrderDT.docstatus == 1)
		.orderby(OrderDT.creation, order=frappe.qb.desc)
		.limit(10)
	)
	return query.run(as_dict=True)


def get_monthly_comparison():
	"""Compare current month vs previous month."""
	OrderDT = DocType("Order")
	today = getdate(nowdate())
	current_month_start = today.replace(day=1)
	prev_month_start = add_months(current_month_start, -1)

	current = (
		frappe.qb.from_(OrderDT)
		.select(
			Count("*").as_("orders"),
			Sum(OrderDT.total_amount).as_("revenue")
		)
		.where(OrderDT.docstatus == 1)
		.where(OrderDT.order_date >= current_month_start)
		.run(as_dict=True)
	)[0]

	previous = (
		frappe.qb.from_(OrderDT)
		.select(
			Count("*").as_("orders"),
			Sum(OrderDT.total_amount).as_("revenue")
		)
		.where(OrderDT.docstatus == 1)
		.where(OrderDT.order_date >= prev_month_start)
		.where(OrderDT.order_date < current_month_start)
		.run(as_dict=True)
	)[0]

	return {
		"current_orders": int(current.orders or 0),
		"current_revenue": float(current.revenue or 0),
		"previous_orders": int(previous.orders or 0),
		"previous_revenue": float(previous.revenue or 0),
	}
